### PR TITLE
Add TPR to CRD migration helper.

### PR DESCRIPTION
**Pull request by enisoc**
This is a helper for migrating TPR data to CustomResource. It's rather hacky because it requires crossing apiserver boundaries, but doing it this way keeps the mess contained to the TPR code, which is scheduled for deletion anyway.

It's also not completely hands-free because making it resilient enough to be completely automated is too involved to be worth it for an alpha-to-beta migration, and would require investing significant effort to fix up soon-to-be-deleted TPR code. Instead, this feature will be documented as a best-effort helper whose results should be verified by hand.

The intended benefit of this over a totally manual process is that it should be possible to copy TPR data into a CRD without having to tear everything down in the middle. The process would look like this:

1. Upgrade to k8s 1.7. Nothing happens to your TPRs.
1. Create CRD with group/version and resource names that match the TPR. Still nothing happens to your TPRs, as the CRD is hidden by the overlapping TPR.
1. Delete the TPR. The TPR data is converted to CustomResource data, and the CRD begins serving at the same REST path.

Note that the old TPR data is left behind by this process, so watchers should not receive DELETE events. This also means the user can revert to the pre-migration state by recreating the TPR definition.

Ref. https://github.com/kubernetes/kubernetes/issues/45728

### CODE REVIEW

File: pkg/master/thirdparty/thirdparty.go [hunk 1]
```diff
@@ -166,25 +180,118 @@ func (m *ThirdPartyResourceServer) RemoveThirdPartyResource(path string) error {
 	return nil
 }
 
-func (m *ThirdPartyResourceServer) removeAllThirdPartyResources(registry *thirdpartyresourcedatastore.REST) error {
+func (m *ThirdPartyResourceServer) removeThirdPartyResourceData(group *metav1.APIGroup, resource string, registry *thirdpartyresourcedatastore.REST) error {
+	// Freeze TPR data to prevent new writes via this apiserver process.
+	// Other apiservers can still write. This is best-effort because there
+	// are worse problems with TPR data than the possibility of going back
+	// in time when migrating to CRD [citation needed].
```
Comment (deads2k) [inline pkg/master/thirdparty/thirdparty.go]: The original issue 95. :)

File: pkg/registry/extensions/thirdpartyresourcedata/storage/storage.go [hunk 2]
```diff
@@ -18,20 +18,63 @@ package storage
...
+
+// REST implements a RESTStorage for ThirdPartyResourceData.
 type REST struct {
 	*genericregistry.Store
-	kind string
+	kind   string
+	frozen atomic.Value
+}
+
+func (r *REST) Freeze() {
```
Comment (deads2k) [inline pkg/registry/extensions/thirdpartyresourcedata/storage/storage.go]: nit: a comment that this is a one way trip.

**Comment by deads2k** [comment]
This is a strong start.

/lgtm

**Comment by deads2k** [comment]
> One thing I haven't handled yet is resetting watchers and preventing them from receiving DELETE events. I also need to add tests, which will likely need to be either e2e or CLI-based. But the core approach can be reviewed now.

What if you detected that this was a migration (you're already doing that) and just didn't delete the original TPR data from etcd?  We'd have dangling entries, but no one would be reading them anymore and no delete events would get sent.  The watch would be dropped and the client's next list/watch would be satisfied by CRD.

File: pkg/master/thirdparty/thirdparty.go [hunk 3]
```diff
@@ -166,25 +180,118 @@ func (m *ThirdPartyResourceServer) RemoveThirdPartyResource(path string) error {
...
+			break
+		}
+	}
+	if crd == nil {
+		// No matching CRD, so no migration necessary.
+		return nil
+	}
+
+	// Talk directly to CustomResource storage.
+	// We have to bypass the API server because TPR is shadowing CRD at this point.
```
Comment (sttts) [inline pkg/master/thirdparty/thirdparty.go]: will a client observe the deletions? Thinking about the etcd operator deleting its storage volumes.

Comment (sttts) [inline pkg/master/thirdparty/thirdparty.go]: If I understand the PR correctly, you delete the TPR and it automatically is migrated (if successful) to the CDR (if it exists). This feels pretty counter-intuitive. I would have expected an start-cdr-migration annotation on the TPR, or even an automatic migration as soon as the CDR shows up.

Comment (sttts) [inline pkg/master/thirdparty/thirdparty.go]: Follow-up bugfix is fine.

**Comment by enisoc** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by enisoc** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by enisoc** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by enisoc** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by sttts** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by nikhita** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Pull request merged**